### PR TITLE
Integrated Killstreak Kit Additions to the script

### PR DESCRIPTION
**Pull request by HairGMan**
I made this small script earlier this year that adds some shortcuts to different killstreak tiers, along with kits/fabricators, to the stats pages of all weapons, kits and fabs. I uploaded it to its own repo last week, then got contacted by Moder112, who lent me a hand with optimizing the code and recommended I make it a part of Misc Extensions, so I did just that! I'm opening this pull request to see if you'd like to add my code to your script.
Here's my script's standalone repo: https://github.com/HairGMan/bptf-killstreak-kit-additions
If you have any questions about the code or you'd want me to change anything, please feel free to tell me! Thanks for your time!

### CODE REVIEW

File: backpack.tf extended sorting.user.js [hunk 1]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+        let splitPath = path.split("/");
+        splitPath[3] = splitPath[3].replace(ksTierDict[currentTier],"");
+        if(tier > 0){
+            splitPath[3] = ksTierDict[tier].concat("",splitPath[3]);
+        }
+        let tieredPath = splitPath.join("/");
+        return tieredPath;
+    }
+
+    class itemData{
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js:1287]: This is a pretty generic name, it would be nice if it would be renamed to something like `ItemStatPageInfo` (or whatever else if you have a better fitting name)

File: backpack.tf extended sorting.user.js [hunk 2]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+        constructor(headerItem){
+            this.name = headerItem.dataset.base_name;
+            this.defindex = headerItem.dataset.defindex;
+            this.killstreak_tier = headerItem.dataset.ks_tier ? headerItem.dataset.ks_tier : 0;
+            this.aussie = headerItem.dataset.australium ? headerItem.dataset.australium : 0;
+            this.slot = headerItem.dataset.slot ? headerItem.dataset.slot : "none";
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: A shorter way to write this expression would be the following
```suggestion
            this.killstreak_tier = headerItem.dataset.ks_tier || 0;
            this.aussie = headerItem.dataset.australium || 0;
            this.slot = headerItem.dataset.slot || "none";
```

File: backpack.tf extended sorting.user.js [hunk 3]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+            1 : "Killstreak%20",
+            2 : "Specialized%20Killstreak%20",
+            3 : "Professional%20Killstreak%20",
+        }
+        let splitPath = path.split("/");
+        splitPath[3] = splitPath[3].replace(ksTierDict[currentTier],"");
+        if(tier > 0){
+            splitPath[3] = ksTierDict[tier].concat("",splitPath[3]);
+        }
+        let tieredPath = splitPath.join("/");
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: (nitpick)
Redundant to declare the variable, you can directly return it.

File: backpack.tf extended sorting.user.js [hunk 4]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+            else{
+                let defindexInt = parseInt(this.defindex);
+                if(this.idreplacedict.hasOwnProperty(defindexInt)){
+                    itemId = this.idreplacedict[defindexInt].toString();
+                }
+                else{
+                    itemId = this.defindex;
+                }
+            }
+            return itemId;
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: (nitpick)
If I don't see it wrongly, you can just return the item id of all your branches directly, instead of storing it in a variable and then returning that variable.

```suggestion
            let itemName = this.name.match(/((?<=((Blood)|(Gold)|(Silver)|(Rust)|(Carbonado)|(Diamond)) Botkiller ).+(?= Mk.II?))|((?<=Festive ).+)/g); // Regex courtesy of Moder112
            if(itemName){
                this.defindex = this.nameiddict[itemName[0]];
                return this.defindex;
            }
            let defindexInt = parseInt(this.defindex);
            if(this.idreplacedict.hasOwnProperty(defindexInt))
                return this.idreplacedict[defindexInt].toString();
            return this.defindex;
```

File: backpack.tf extended sorting.user.js [hunk 5]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+                    else if(this.name == "Fabricator"){
+                        itemSlot = -2;
+                    }
+                    else{
+                        itemSlot = 0;
+                    }
+                    break;
+            }
+            return itemSlot;
+        }
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: (nitpick first)
As before, you can directly return the value of your switch case instead of storing it in a variable, then you also wouldn't need the break (nor the else condition)

---------------------------------------------

I think it would be clearer to return a string instead of a number which you later use in the switch statement, as this would be clearer for the reader of the code.

(For example "weapon", "misc", "kit", "fabricator" instead of 1,0,-1,-2)


Would look the following:

```suggestion
        getItemSlot(){
            switch(this.slot){
                case "primary": case "secondary": case "melee": case "pda":
                    return "weapon"
                case "misc":
                    return "misc"
                case "none":
                    if(this.name == "Kit")
                        return "kit"
                    if (this.name == "Fabricator") 
                        return "fabricator"
                    return "misc"
            }
        }
```

File: backpack.tf extended sorting.user.js [hunk 6]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+
+        getKillstreakTier(){
+            return this.killstreak_tier;
+        }
+    };
+
+    if (window.location.pathname.startsWith("/stats")) {
+        const currentItem = new itemData(document.querySelector(".item"));
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: itemData would need to be renamed to the new class name

File: backpack.tf extended sorting.user.js [hunk 7]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+        let currentPath = window.location.pathname;
+
+        $(".stats-header-controls").append(`
+     <br>
+     <div class="btn-group btn-group-sm stats-killstreak-list" style="margin: 0 0 12px"></div>
+     `);
+
+        switch(currentItem.getItemSlot()){
+
+            case 0:
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]:
```suggestion
            case "misc":
```

File: backpack.tf extended sorting.user.js [hunk 8]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+                break;
+
+            case -1: {
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]:
```suggestion
            case "kit": {
```

File: backpack.tf extended sorting.user.js [hunk 9]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+        } else {
+            $parent.addClass('expanded');
+        }
+    });
+</script>`);
+                });
+                break;
+            }
+
+            case -2: {
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]:
```suggestion
            case "fabricator": {
```

File: backpack.tf extended sorting.user.js [hunk 10]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+                $(".stats-header-controls").append(`
+     <a class="btn btn-default" href="https://backpack.tf/stats/Unique/Killstreak%20Kit/Tradable/Non-Craftable/${killstreakTier}-${weaponDefIndex}" style="margin: 0px 0px 12px;">
+     <i class="fa fa-bar-chart"></i>
+     Kit Stats
+     </div>
+     `);
+                break;
+            }
+
+            case 1: {
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]:
```suggestion
            case "weapon": {
```

File: backpack.tf extended sorting.user.js [hunk 11]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+     `);
+
+        switch(currentItem.getItemSlot()){
+
+            case 0:
+                break;
+
+            case -1: {
+
+                let weaponName = trimSpaces(document.querySelector(".stats-breadcrumbs").children[1].innerHTML);
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: A better way would be here to use the following instead of your method:
```suggestion
                let weaponName = encodeURIComponent($(".stats-breadcrumb").last().text().trim())
```

(You can keep your pure DOM Manipulation if you want, but JavaScript already has a method for trimming, and instead of manually replacing the space character with %20 it would be best to use the JavaScript function to encode URI components (which does exactly that among others))

File: backpack.tf extended sorting.user.js [hunk 12]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+                let weaponDefIndex = window.location.pathname.match(/\d+$/g);
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: This method works perfectly fine, but it would additionally possible to use:

```suggestion
                let weaponDefIndex = $(".stats-header-item .item").attr("data-priceindex").split("-").pop()
```
-------------------------------
Exactly the same code (`$(".stats-header-item .item").attr("data-priceindex").split("-").pop()`) would work for fabricators too.

File: backpack.tf extended sorting.user.js [hunk 13]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+     <a class="btn btn-variety q-440-text-7" href="https://backpack.tf/stats/Unique/Killstreak%20Kit/Tradable/Non-Craftable/2-${weaponDefIndex}"> Specialized </a>
+     <a class="btn btn-variety q-440-text-11" href="https://backpack.tf/stats/Unique/Killstreak%20Kit/Tradable/Non-Craftable/3-${weaponDefIndex}"> Professional </a>
+     `);
+                document.querySelector(".stats-killstreak-list").children[killstreakTier-1].classList.add("active");
+                if(killstreakTier > 1){
+                    $(".stats-header-controls").append(`
+     <a class="btn btn-default" href="https://backpack.tf/stats/Unique/Fabricator/Tradable/Craftable/${6520 + 3 * (killstreakTier - 1)}-6-${weaponDefIndex}" style="margin: 0px 0px 12px;">
+     <i class="fa fa-bar-chart"></i>
+     Fabricator Stats
+     </div>
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]:
```suggestion
     </a>
```

File: backpack.tf extended sorting.user.js [hunk 14]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+    $('#btn-expand-list').click(function () {
+        var $parent = $(this).parent();
+
+        if ($parent.hasClass('expanded')) {
+            $parent.removeClass('expanded');
+        } else {
+            $parent.addClass('expanded');
+        }
+    });
+</script>`);
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: You could shorten this to the following:

```suggestion
                for(let button of this.firstChild.children){
                    button.href = appendKSToPath(killstreakTier, button.getAttribute("href"), 0);
                }
		$(this.firstChild).removeClass("overview-quality-list expanded")
			.append(`<a class="btn btn-variety" id="btn-expand-list"><i class="fa fa-ellipsis-h"></i></a>`);

		  $('#btn-expand-list').click(function () {
			  $(this).parent().toggleClass("expanded");
		  });
```

(Use JQuery to remove the classes and chain the methods, additionally use an JQuery Event to toggle the class, instead of adding an additional script tag to remove and add classes by hand) 

If I am not wrong removing the outer container class should be redundant (style wise) so I removed it, but you can also keep it.

File: backpack.tf extended sorting.user.js [hunk 15]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
 		childList: true
 	});
 
+    // ################# Killstreak Kit Additions for Stats pages ##############
+
+    function trimSpaces(string){
+        return string.replace(/^\s+/g, '').replace(/\s+$/g, '').replace(/\s+/g,"%20");
+    }
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: This function is redundant, a better alternative is given below

File: backpack.tf extended sorting.user.js [hunk 16]
```diff
@@ -1268,4 +1268,246 @@ class</a></li>
...
+            }
+        }
+
+        getItemName(){
+            let itemName = this.name;
+            if(this.aussie == 1){
+                itemName = "Australium".concat(" ", itemName);
+            }
+            return itemName;
+        }
```
Comment (NetroScript) [inline backpack.tf extended sorting.user.js]: (nitpick)
Why not just:

```suggestion
        getItemName(){
            return (this.aussie == 1) ? "Australium " + this.name : this.name;
        }
```

?

**Review by NetroScript — CHANGES_REQUESTED** [review]
Thank you for your Pull Request!

This is a worthy addition to this userscript.

There are some small things I would change, I commented on them here.

Don't be surprised about the many comments, your code is working perfectly fine, I would just change some small things and I did some small suggestions how I would change it considering readability and shortening code.

File: backpack.tf extended sorting.user.js (hunk 13 above)
Comment (NetroScript) [reply backpack.tf extended sorting.user.js]: You open an anchor tag, but close a div tag, the same is further below too if I remember correctly, I just forgot to make a suggestion.

**Comment by NetroScript** [comment]
Ah and one more thing I forgot to mention is, that you also should increment the version of the script to 0.1.26 (don't forget the header too) and also add the changes to the changelog in the `README`. Since 0.1.25 your functionality has been added (after merging this) and additionally:

```
Fixed:

* Incorrect classifieds queries on paintkit items
```

which was the previous commit by @Moder112

**Comment by HairGMan** [comment]
Thank you so much for the review! I already commited all the changes you suggested.
This is the first userscript I've ever written, so all the corrections, even the small things, are appreciated, and I'll keep them in mind for the future! If you need anything else changed before merging, hit me up and I'll add it.

**Review by NetroScript — APPROVED** [review]

**Comment by NetroScript** [comment]
Thank you again for your time and the PR!
